system-parachains/coretime/coretime-kusama/src/coretime.rs

```diff
@@ -0,0 +1,249 @@
+}
+
+/// Type that implements the `CoretimeInterface` for the allocation of Coretime. Meant to operate
+/// from the parachain context. That is, the parachain provides a market (broker) for the sale of
+/// coretime, but assumes a `CoretimeProvider` (i.e. a Relay Chain) to actually provide cores.
+pub struct CoretimeAllocator;
+impl CoretimeInterface for CoretimeAllocator {
+	type AccountId = AccountId;
+	type Balance = Balance;
+	type RealyChainBlockNumberProvider = RelaychainDataProvider<Runtime>;
+
+	fn request_core_count(count: CoreIndex) {
+		use crate::coretime::CoretimeProviderCalls::RequestCoreCount;
+		let request_core_count_call = RelayRuntimePallets::Coretime(RequestCoreCount(count));
+
+		// Weight for `request_core_count` from Kusama runtime benchmarks:
+		// `ref_time` = 7889000 + (3 * 25000000) + (1 * 100000000) = 182889000
+		// `proof_size` = 1636
+		// Add 5% to each component and round to 2 significant figures.
+		let call_weight = Weight::from_parts(190_000_000, 1700);
+
+		let message = Xcm(vec![
+			Instruction::UnpaidExecution {
+				weight_limit: WeightLimit::Unlimited,
+				check_origin: None,
+			},
+			Instruction::Transact {
+				origin_kind: OriginKind::Native,
+				require_weight_at_most: call_weight,
+				call: request_core_count_call.encode().into(),
+			},
+		]);
+
+		match PolkadotXcm::send_xcm(Here, Location::parent(), message) {
+			Ok(_) => log::debug!(
+				target: "runtime::coretime",
+				"Request to update schedulable cores sent successfully."
+			),
+			Err(e) => log::error!(
+				target: "runtime::coretime",
+				"Failed to send request to update schedulable cores: {:?}",
+				e
+			),
+		}
+	}
+
+	fn request_revenue_info_at(when: RCBlockNumberOf<Self>) {
+		use crate::coretime::CoretimeProviderCalls::RequestRevenueInfoAt;
+		let _request_revenue_info_at_call =
+			RelayRuntimePallets::Coretime(RequestRevenueInfoAt(when));
+
+		log::debug!(
+			target: "runtime::coretime",
+			"`request_revenue` is unmiplemented on the relay."
+		);
+	}
+
+	fn credit_account(who: Self::AccountId, amount: Self::Balance) {
+		use crate::coretime::CoretimeProviderCalls::CreditAccount;
+		let _credit_account_call = RelayRuntimePallets::Coretime(CreditAccount(who, amount));
+
+		log::debug!(
+			target: "runtime::coretime",
+			"`credit_account` is unmiplemented on the relay."
+		);
+	}
+
+	fn assign_core(
+		core: CoreIndex,
+		begin: RCBlockNumberOf<Self>,
+		assignment: Vec<(CoreAssignment, PartsOf57600)>,
+		end_hint: Option<RCBlockNumberOf<Self>>,
+	) {
+		use crate::coretime::CoretimeProviderCalls::AssignCore;
+		let assign_core_call =
+			RelayRuntimePallets::Coretime(AssignCore(core, begin, assignment, end_hint));
+
+		// Weight for `assign_core` from Kusama runtime benchmarks:
+		// `ref_time` = 10177115 + (1 * 25000000) + (2 * 100000000) + (80 * 13932) = 236291675
+		// `proof_size` = 3612
+		// Add 5% to each component and round to 2 significant figures.
```

Comment:
Do we need to be so cheap? Going to high here only limits how many messages we can send - right? I think a safety margin of e.g. 50% would add significantly more robustness against code upgrades.

Reply:
I had wondered about this, my reasoning was the following: for `assign_core` we send one of these messages for each core every timeslice, so it's relatively often - but yes only affects the number we can send each time. I didn't work out how many would fit in a block, but since the worst case is 80 assignments per call this is already quite a lot more than the median expected (obviously we don't have numbers for this, but I think it's reasonable to expect that quite a lot of these will be a maximum of a handful of assignments). So really the 5% buffer is 5% on top of the worst case benchmark for the worst case core, but over 100% buffer in the best case of one assignment per core.

That’s for the ref time only, the proof size really is just 5% as it doesn't depend on the number of assignments per core.

I guess we could factor out the part that depends on the number of assignments and multiply that by the length of the actual assignments `Vec` that comes in and slap 50% on that?

Reply:
> core every timeslice

Is this true? If assignments don't change, this is actually not required.

> That’s for the ref time only, the proof size really is just 5% as it doesn't depend on the number of assignments per core.

Why is that? It should, shouldn't it?

Anyhow, your argument that this is already 5% on the absolute worst case, and that we might end up having a lot of these messages makes sense. Any hardcoded value will be brittle anyway.